pkg/tsdb/grafana-testdata-datasource/admission_handler.go

```diff
@@ -0,0 +1,78 @@
+package testdatasource
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ValidateAdmission implements backend.AdmissionHandler.
+func (s *Service) ValidateAdmission(ctx context.Context, req *backend.AdmissionRequest) (*backend.AdmissionResponse, error) {
+	expected := (&backend.DataSourceInstanceSettings{}).GVK()
+	if req.Kind.Kind != expected.Kind && req.Kind.Group != expected.Group {
+		return getBadRequest("expected DataSourceInstanceSettings protobuf payload"), nil
+	}
+
+	// Convert the payload from protobuf to an SDK struct
+	settings, err := backend.DataSourceInstanceSettingsFromProto(req.ObjectBytes, "")
+	if err != nil {
+		return nil, err
+	}
+	if settings == nil {
+		return getBadRequest("missing datasource settings"), nil
+	}
+
+	switch settings.APIVersion {
+	case "", "v0alpha1":
+		// OK!
+	default:
+		return getBadRequest(fmt.Sprintf("expected apiVersion: v0alpha1, got: %s", settings.APIVersion)), nil
+	}
+	if settings.JSONData != nil {
+		anything := map[string]any{}
+		err := json.Unmarshal(settings.JSONData, &anything)
+		if err != nil || len(anything) > 0 {
+			return getBadRequest("Expected empty jsonData settings"), nil
+		}
+	}
+	if len(settings.DecryptedSecureJSONData) > 0 {
+		return getBadRequest("found unsupported secure json fields"), nil
+	}
+	if settings.URL != "" {
+		return getBadRequest("unsupported URL value"), nil
+	}
+	if settings.User != "" {
+		return getBadRequest("unsupported User value"), nil
+	}
+	return &backend.AdmissionResponse{Allowed: true}, nil
+}
+
+// MutateAdmission implements backend.AdmissionHandler.
+func (s *Service) MutateAdmission(ctx context.Context, req *backend.AdmissionRequest) (*backend.AdmissionResponse, error) {
+	rsp, err := s.ValidateAdmission(ctx, req)
+	if err != nil && rsp.Allowed {
+		rsp.ObjectBytes = req.ObjectBytes // For this case they are the same!
+	}
```

Comment:
isn't this a hack because Grafana doesn't call the validation function, only the mutation one?

Reply:
ya -- for the test data case, both are equivalent.  A valid config is empty.

For other cases -- they may be different.  This for example is where you can accept old formats, but make sure you save one structured how you want.